profiling/src/collections/string_storage.rs

```diff
@@ -0,0 +1,148 @@
+            next_id: 0,
+            id_to_data: Default::default(),
+            str_to_id: Default::default(),
+            current_gen: 0,
+        };
+        // Ensure empty string gets id 0 and always has usage > 0 so it's always retained
+        storage.intern("");
+        storage
+    }
+
+    pub fn advance_gen(&mut self) {
+        self.id_to_data.retain(|_, data| {
+            /*let used_in_this_gen = data
+                .last_usage_gen
+                .get()
+                .map_or(false, |last_usage_gen| last_usage_gen == self.current_gen);
+            if used_in_this_gen || *id == 0 {
+                // Empty string (id = 0) or anything that was used in the gen
+                // we are now closing, is kept alive
+                return true;
+            }*/
+            if data.usage_count.get() > 0 {
+                return true;
+            }
+
+            self.str_to_id.remove_entry(&data.str);
+            false
+        });
+        self.current_gen += 1;
+    }
+
+    pub fn intern(&mut self, item: &str) -> u32 {
+        let entry = self.str_to_id.get_key_value(item);
+        match entry {
+            Some((_, id)) => {
+                let usage_count = &self
+                    .id_to_data
+                    .get(id)
+                    .expect("id_to_data and str_to_id should be in sync")
+                    .usage_count;
+                usage_count.set(usage_count.get() + 1);
+                *id
+            }
+            None => {
+                let id = self.next_id;
+                let str: Rc<str> = item.into();
+                let data = ManagedStringData {
+                    str: str.clone(),
+                    last_usage_gen: Cell::new(None),
+                    cached_seq_num_for: Cell::new(None),
+                    cached_seq_num: Cell::new(None),
+                    usage_count: Cell::new(1),
+                };
+                self.next_id = self.next_id.checked_add(1).expect("Ran out of string ids!");
+                let old_value = self.str_to_id.insert(str.clone(), id);
+                debug_assert_eq!(old_value, None);
+                let old_value = self.id_to_data.insert(id, data);
+                debug_assert_eq!(old_value, None);
+                id
+            }
+        }
+    }
+
+    pub fn unintern(&self, id: u32) {
+        if id == 0 {
+            panic!("For performance reasons, unintern should not be called with id == 0. Please hardcode a fast path check in the caller to avoid this call")
+        }
+
+        let data = self.get_data(id);
+        let usage_count = &data.usage_count;
+        usage_count.set(usage_count.get() - 1);
+    }
+
+    pub fn get_seq_num(&self, id: u32, profile_strings: &mut StringTable) -> StringId {
+        if id == 0 {
+            panic!("For performance reasons, get_set_num should not be called with id == 0. Please hardcode a fast path check in the caller to avoid this call")
+        }
```

Comment:
This panic shouldn't exist either, right? I think avoiding it in the caller is probably okay since it's a `0` to `0` mapping without side effects, though.

Reply:
This is here for the same reason as above: not for correctness, but for performance -- callers must do their own checks for zero rather then letting it get to the string table.

This being rust, we can probably in the future be a bit fancier and encode this in a type that can never be zero ;)